### ISSUE TITLE
Should p4c give error for programs using direct type invocation more than once in same scope?

### ISSUE THREAD

**Issue by jafingerhut** [p4lang/p4c#2203]
I believe that this v1model architecture P4_16 program uses direct type invocation multiple times in the same scope (of the ingress control), on the control named `tableControl`: https://github.com/jafingerhut/p4-guide/blob/master/instantiation/table-ctrl-dti-2-times.p4#L50-L86

The P4_16 language spec, version 1.2.0, Section 14.1 "Direct type invocation", says the following:

"In the same scope, direct type invocation will result in a different local instance per invocation -- however, instances of the same type will share the same global name, via the @name annotation. If the type contains controllable entities, then invoking it directly more than once in the same scope is illegal, because it will produce multiple controllable entities with the same fully-qualified control name."

Questions:

+ Is the linked P4 program using direct type invocation more than once in the same scope?  I think so, but want to verify my understanding.
+ Is the `tableControl` control an example of a type that contains controllable entities?  I think it is, because it contains a table, which is one of the controllable entities listed in Section 17.3 "Control plane names".
+ If the answer to the above two questions is yes for both, is this program illegal according to the language spec?
+ If it is illegal, should p4c give an error for it?  The latest version of p4c as of 2020-Feb-18 (commit 62fd40a0097b361c266ede5116ac0060d1b27406) gives no error for any of the following commands, and for the last one it writes a BMv2 JSON file:

```
$ p4test table-ctrl-dti-2-times.p4 
$ p4c table-ctrl-dti-2-times.p4 
$ p4c --target bmv2 --arch v1model table-ctrl-dti-2-times.p4 
```

On the bright side, trying to run `simple_switch` on the BMv2 JSON file produced does give a pretty useful error message:
```
$ simple_switch table-ctrl-dti-2-times.json
Duplicate objects of type 'control node' with name 'ingressImpl.tableControl.foo'
```
However, if the program is illegal, I wonder whether it might be worth p4c catching this all by itself, versus producing a compiled output with an error caught by a later program.

**Comment by mbudiu-vmw** [comment]
I think that if you try to generate the control-plane API you will get an error.

**Comment by jafingerhut** [comment]
Yes, attempting to generate the P4Info file using the `--p4runtime-files` command line option does result in an error message from the latest `p4c`:
```
$ p4c --target bmv2 --arch v1model --p4runtime-files foo.p4info.txt table-ctrl-dti-2-times.p4 
[--Werror=duplicate] error: Name 'ingressImpl.tableControl.foo' is used for multiple table objects in the P4Info message
[--Werror=duplicate] error: Found 1 duplicate name(s) in the P4Info
```

If you are suggesting that we recommend to people that they use the `--p4runtime-files` option in order to catch such errors, that seems like one viable approach.

**Comment by mbudiu-vmw** [comment]
These duplicate names are only meaningful for the control-plane. Whatever method is used to generate the control-plane API should perform these checks. If no control-plane API is needed (for some strange reason), then perhaps there is no need to give an error at all?
